src/module_3/model_selection_converted.md

```diff
@@ -0,0 +1,524 @@
+
+
+```python
+daily_orders = final_data.groupby('order_date').count()
+plt.figure(figsize=(10, 6)) 
+plt.plot(daily_orders, marker='o', linestyle='-')  
+plt.title('Daily Orders')  
+plt.xlabel('Date')  
+plt.ylabel('Number of Orders')
+plt.grid(True) 
+plt.show()
+```
+
+Firstly, we should conduct research regarding date information:
+
+
+```python
+print("Minimum order date:", final_data['order_date'].min())
+print("Maximum order date:", final_data['order_date'].max())
+```
+
+The criterion followed for the separation is based on dividing the data into three sets: training, validation, and testing, while maintaining a temporal relationship among them. This approach ensures that the model is trained on data from an earlier period than the validation and testing data.
+
+Specifically, the data is divided into three parts using quantiles based on the order dates. Approximately 70% of the data is allocated to the training set, while approximately 15% each is allocated to the validation and testing sets. This allocation strategy considers the distribution of orders over time, aiming to prevent information contamination between sets.
+
+Adhering to this temporal relationship is crucial for preventing the model from learning patterns specific to the validation and testing data, thus promoting its generalization ability to unseen data effectively. Moreover, allocating a larger portion of data to training ensures that the model learns from a diverse range of examples, which enhances its overall performance and robustness.
+
+
+```python
+final_data_sorted = final_data.sort_values(by='order_date')
+
+train_end = final_data_sorted['order_date'].quantile(0.7)
+val_end = final_data_sorted['order_date'].quantile(0.85)
+test_end = final_data_sorted['order_date'].quantile(1)
+
+
+train_data = final_data_sorted[final_data_sorted['order_date'] < train_end]
+val_data = final_data_sorted[(final_data_sorted['order_date'] >= train_end) & (final_data_sorted['order_date'] < val_end)]
+test_data = final_data_sorted[(final_data_sorted['order_date'] >= val_end) & (final_data_sorted['order_date'] <= test_end)]
+
+print("Training set size:", (train_data.shape[0] / final_data_sorted.shape[0]))
+print("Validation set size:", (val_data.shape[0] / final_data_sorted.shape[0]))
+print("Testing set size:", (test_data.shape[0] / final_data_sorted.shape[0]))
+```
+
+These dates represent the end points of each subset:
+
+* The training set ends on February 9, 2021, at midnight.
+* The validation set ends on February 21, 2021, at midnight.
+* The testing set ends on March 3, 2021, at midnight.
+
+**Possible issue**: What if I have a user who makes purchases in both the training and testing sets? Will the model learn from that user? Is there a risk of data leakage?
```

Comment:
Nope, because that's what will happen in prod.

Reply:
Anyway, a good question would be to evaluate how a model generalise to users not included in training set due to sampling, for example.